File: nameban.go
package unrealircd

import (
	"errors"
)

// NameBan handles name ban (QLine) operations
type NameBan struct {
	querier Querier
}

// Add adds a name ban (QLine)
func (nb *NameBan) Add(name, reason string, duration, setBy *string) (interface{}, error) {
	params := map[string]interface{}{
		"name":    name,
		"reason":  reason,
		"duration_string": "0",
	}
	if duration != nil {
		params["duration_string"] = *duration
	}
	if setBy != nil {
		params["set_by"] = *setBy
	}

	result, err := nb.querier.Query("name_ban.add", params, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if tkl, ok := res["tkl"]; ok {
			return tkl, nil
		}
	}

	return nil, nil
}

// Delete deletes a ban
func (nb *NameBan) Delete(name string) (interface{}, error) {
	result, err := nb.querier.Query("name_ban.del", map[string]interface{}{
		"name": name,
	}, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if tkl, ok := res["tkl"]; ok {
			return tkl, nil
		}
	}

	return nil, nil
}

// GetAll returns a list of all bans
func (nb *NameBan) GetAll() (interface{}, error) {
	result, err := nb.querier.Query("name_ban.list", nil, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if list, ok := res["list"]; ok {
			return list, nil
		}
	}

	return nil, errors.New("Invalid JSON Response from UnrealIRCd RPC")
}

// Get gets a specific ban
func (nb *NameBan) Get(name string) (interface{}, error) {
	result, err := nb.querier.Query("name_ban.get", map[string]interface{}{
		"name": name,
	}, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if tkl, ok := res["tkl"]; ok {
			return tkl, nil
		}
	}

	return nil, nil // not found
}

File: UnrealIRCdRPC/NameBan.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Text.Json;

namespace UnrealIRCdRPC
{
    // NameBan handles name ban operations
    public class NameBan
    {
        private readonly IQuerier _querier;

        public NameBan(IQuerier querier)
        {
            _querier = querier;
        }

        // Add adds a name ban
        public async Task<JsonElement?> AddAsync(string name, string reason, string? duration = null, string? setBy = null)
        {
            var parameters = new Dictionary<string, object>
            {
                { "name", name },
                { "reason", reason }
            };
            if (duration != null) parameters["duration_string"] = duration;
            if (setBy != null) parameters["set_by"] = setBy;
            var result = await _querier.QueryAsync("name_ban.add", parameters, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object &&
                result.Value.TryGetProperty("tkl", out var tklElement))
            {
                return tklElement;
            }
            return null;
        }

        // Delete deletes a name ban
        public async Task<JsonElement?> DeleteAsync(string name)
        {
            var result = await _querier.QueryAsync("name_ban.del", new { name }, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object &&
                result.Value.TryGetProperty("tkl", out var tklElement))
            {
                return tklElement;
            }
            return null;
        }

        // GetAll returns a list of all name bans
        public async Task<JsonElement?> GetAllAsync()
        {
            var result = await _querier.QueryAsync("name_ban.list", null, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object &&
                result.Value.TryGetProperty("list", out var listElement))
            {
                return listElement;
            }
            throw new Exception("Invalid JSON Response from UnrealIRCd RPC");
        }

        // Get gets a specific name ban
        public async Task<JsonElement?> GetAsync(string name)
        {
            var result = await _querier.QueryAsync("name_ban.get", new { name }, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object &&
                result.Value.TryGetProperty("tkl", out var tklElement))
            {
                return tklElement;
            }
            return null; // not found
        }
    }
}